src/uri/encoding.cr

```diff
@@ -347,4 +376,11 @@ class URI
     i += 1
     i
   end
+
+  # Checks if there's any chars that actually needs decoding.
+  # If not, decoding a string can be done much faster, because
+  # it's just the same string.
+  private def self.needs_decode?(string : String, *, plus_to_space : Bool) : Bool
+    !!(string.index('%') || (plus_to_space && string.index('+')))
```

Comment:
Can this be simplified to 

```suggestion
    string.includes?('%') || (plus_to_space && string.includes?('+')))
```

`includes?` is defined as `!!index(search)`, so it's the same, but `includes?` is more readable IMO

Reply:
Nice, thank you!

Reply:
What about something like:

```cr
private def needs_decode_any?(string : String, *, plus_to_space : Bool) : Bool
  string.each_char do |chr|
    return true if chr == '%' || (plus_to_space && chr == '+')
  end

  false
end
```

I think that would avoid iterating over the whole string possibly twice.

Reply:
That was my first code. The thing is that using `memchr` is faster because `memchr` is extremely optimized. I think `memchr` doesn't go one by one, it goes by larger chunks and does some bit magic. No idea, to be honest. But going char by char or byte by byte, even if we only do it once, is slower.

Reply:
@asterite Interesting, I did a quick benchmark and that deff seems to be the case on the no decoding side of things:

<details>

private def needs_decode?(string : String, *, plus_to_space : Bool) : Bool
  string.includes?('%') || (plus_to_space && string.includes?('+'))
end

private def needs_decode_iterate?(string : String, *, plus_to_space : Bool) : Bool
  string.each_char do |chr|
    return true if chr == '%' || (plus_to_space && chr == '+')
  end

  false
end

require "benchmark"

Benchmark.ips do |x|
  x.report "no decode current" do
    needs_decode? "/foobar", plus_to_space: false
  end

  x.report "no decode new" do
    needs_decode_iterate? "/foobar", plus_to_space: false
  end
end

puts
puts

Benchmark.ips do |x|
  x.report "decode current" do
    needs_decode? "/foo%20bar", plus_to_space: false
  end

  x.report "decode new" do
    needs_decode_iterate? "/foo%20bar", plus_to_space: false
  end
end

puts
puts

Benchmark.ips do |x|
  x.report "decode current plus" do
    needs_decode? "/foo+bar", plus_to_space: true
  end

  x.report "decode new plus" do
    needs_decode_iterate? "/foo+bar", plus_to_space: true
  end
end

# no decode current 143.92M (  6.95ns) (± 6.60%)  0.0B/op        fastest
#     no decode new 127.34M (  7.85ns) (± 5.98%)  0.0B/op   1.13× slower

# decode current 126.47M (  7.91ns) (± 5.60%)  0.0B/op   1.51× slower
#     decode new 191.53M (  5.22ns) (± 5.00%)  0.0B/op        fastest

# decode current plus  75.46M ( 13.25ns) (± 4.39%)  0.0B/op   2.28× slower
#     decode new plus 172.08M (  5.81ns) (± 5.72%)  0.0B/op        fastest

</details>

Not saying we should change it, just noting my findings for future reference.

Reply:
@asterite The diff would just be replacing the current `needs_decode?` with the one from https://github.com/crystal-lang/crystal/pull/11741#discussion_r783199424.

Reply:
Strange. This is what I see on my side...

Current PR:

```
     decode without escapes 102.50M (  9.76ns) (± 9.03%)  0.0B/op        fastest
         decode with escape   7.09M (141.00ns) (± 3.94%)  208B/op  14.45× slower
decode without escapes (IO)  58.97M ( 16.96ns) (± 5.93%)  0.0B/op   1.74× slower
    decode with escape (IO)  14.51M ( 68.92ns) (± 5.73%)  0.0B/op   7.06× slower
```

With your suggestion:

```
     decode without escapes  66.11M ( 15.13ns) (±12.91%)  0.0B/op        fastest
         decode with escape   5.59M (178.82ns) (± 7.54%)  208B/op  11.82× slower
decode without escapes (IO)  38.72M ( 25.83ns) (± 5.62%)  0.0B/op   1.71× slower
    decode with escape (IO)  11.91M ( 83.99ns) (± 8.48%)  0.0B/op   5.55× slower
```

All cases are slower.

Did you benchmark `URI.decode` or just the difference between the two methods?

Reply:
Yea it was just between the two methods. Guess there's some magic going on as you said, so this looks good to me! thanks!

Reply:
It might be worth a try to pull `plus_to_space` out of the equation. The branching really doesn't need to happen in every iteration.

```cr
if plus_to_space
  string.each_char do |chr|
    return true if chr == '%' || chr == '+'
  end
else
  string.includes?('%')
end
```
I don't expect any benefit over the current code because of the high optimization of `memchr`, though. It could provide an improvement for similar use cases, though.

Reply:
I wrote this benchmark:

```crystal
require "uri"
require "benchmark"

io = IO::Memory.new

Benchmark.ips do |x|
  x.report("decode without escapes") do
    URI.decode("hello abc world")
  end
  x.report("decode with escape") do
    URI.decode("hello world %31")
  end
  x.report("decode without escapes (IO)") do
    URI.decode("hello abc world", io)
    io.clear
  end
  x.report("decode with escape (IO)") do
    URI.decode("hello world %31", io)
    io.clear
  end
  x.report("decode without escapes (plus_to_space)") do
    URI.decode("hello abc world", plus_to_space: true)
  end
  x.report("decode with escape (plus_to_space)") do
    URI.decode("hello world +", plus_to_space: true)
  end
  x.report("decode without escapes (IO) (plus_to_space)") do
    URI.decode("hello abc world", io, plus_to_space: true)
    io.clear
  end
  x.report("decode with escape (IO) (plus_to_space)") do
    URI.decode("hello world +", io)
    io.clear
  end
end
```

Results with this PR:

```
                     decode without escapes 128.50M (  7.78ns) (± 8.85%)  0.0B/op        fastest
                         decode with escape   6.82M (146.54ns) (± 3.48%)  208B/op  18.83× slower
                decode without escapes (IO)  63.57M ( 15.73ns) (± 5.12%)  0.0B/op   2.02× slower
                    decode with escape (IO)  15.71M ( 63.64ns) (± 6.97%)  0.0B/op   8.18× slower
     decode without escapes (plus_to_space)  70.87M ( 14.11ns) (± 6.85%)  0.0B/op   1.81× slower
         decode with escape (plus_to_space)   6.88M (145.39ns) (± 4.88%)  208B/op  18.68× slower
decode without escapes (IO) (plus_to_space)  45.06M ( 22.19ns) (± 6.68%)  0.0B/op   2.85× slower
    decode with escape (IO) (plus_to_space)  60.82M ( 16.44ns) (± 5.96%)  0.0B/op   2.11× slower
```

Results with the proposed change:

```
                     decode without escapes 154.99M (  6.45ns) (± 4.01%)  0.0B/op        fastest
                         decode with escape   7.17M (139.40ns) (± 1.91%)  208B/op  21.61× slower
                decode without escapes (IO)  60.90M ( 16.42ns) (± 3.26%)  0.0B/op   2.54× slower
                    decode with escape (IO)  18.15M ( 55.08ns) (± 3.82%)  0.0B/op   8.54× slower
     decode without escapes (plus_to_space)  57.21M ( 17.48ns) (± 3.43%)  0.0B/op   2.71× slower
         decode with escape (plus_to_space)   6.68M (149.65ns) (± 1.21%)  208B/op  23.19× slower
decode without escapes (IO) (plus_to_space)  39.84M ( 25.10ns) (± 2.30%)  0.0B/op   3.89× slower
    decode with escape (IO) (plus_to_space)  58.19M ( 17.19ns) (± 3.51%)  0.0B/op   2.66× slower
```

Note that the four first cases seem to be faster, but that's maybe because my machine was busier in the previous run? But notice the four last cases: they are slower. So my guess is that using `each_char` is always slower than two `memchr`. Remember, `memchr` is usually highly optimized assembly code.

That said, maybe the results depend on the platform. I'm on Mac.

Finally, optimizing this method seems to be a micro-optimization to me, so maybe going deeper might not be worth it. But this entire PR is a big optimization.

Reply:
Oh, I forgot a `plus_to_space: true` in the last one... but I don't have more time for this today. This can always be further improved in a later PR 🙏